apps/aering/src/aer_icode_to_asm.erl

```diff
@@ -0,0 +1,63 @@
+%%%-------------------------------------------------------------------
+%%% @author Happi (Erik Stenman)
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc
+%%%     Translator from Aering Icode to Aevm Assebly
+%%% @end
+%%% Created : 21 Dec 2017
+%%% 
+%%%-------------------------------------------------------------------
+-module(aer_icode_to_asm).
+
+-export([convert/2]).
+
+-include_lib("aebytecode/include/aeb_opcodes.hrl").
+-include("aer_icode.hrl").
+
+convert(#{ contract_name := Name
+         , functions := Functions
+              },
+        Options) ->
+    {_Env, Code} =
+        assemble(Functions,  #{ options => Options
+                              , vars => []
+                              , functions => []
+                              , sp => 0
+                              , frame_size => 0
+                              , address => 0
+                              },
+                 [aeb_opcodes:mnemonic(?COMMENT), "CONTRACT: " ++ Name]),
+    Code.
+
+assemble([{Name, Args, Body}|More], #{frame_size := FS
+                                     , address := Address} = Env, Code) ->
+    Env0 = add_function(Name, Address, Env),
+    {BodyCode, NewEnv} = assemble_body(Body, add_args(Args, Env0, FS), Code),
+    assemble(More, NewEnv, BodyCode);
+assemble([], Env, Code) -> {Env, Code}.
+
+
+assemble_body({var_ref, Id}, #{sp := SP} = Env, Code) ->
+    SL = lookup_var(Id, Env),
+    Instr = aeb_opcodes:mnemonic(?DUP1 + SP + SL),
```

Comment:
Can we have a function to compute a suitable `?DUP` (and others) opcode, so that the knowledge that `?DUP2 = ?DUP1 + 1` is more contained?

Reply:
Ill make a ticket.